Migrate websocket integration tests over to using mint_web_socket library

electricshaman opened the pull request:
Per https://github.com/phoenixframework/phoenix/issues/4699 👏 @the-mikedavis

I think it makes sense to start here with a dependency on `mint_web_socket` at `0.2.0` so there's an opportunity to catch any minor issues which could be addressed in the forthcoming `1.0` release of `mint_web_socket`.

At the moment, I've left the `websocket_client` dependency in place to keep builds stable.  The client API is likely to change anyway due to the style differences (functional vs. process-oriented) already noted between Mint.WebSocket and `websocket_client`.

electricshaman committed:
Add mint_web_socket dependency in test only

electricshaman committed:
Take explicit test dependency on Mint for Mint.HTTP.t

the-mikedavis commented on mix.exs:
```diff
@@ -91,6 +91,8 @@ defmodule Phoenix.MixProject do
       {:phoenix_html, "~> 3.0", only: [:docs, :test]},
       {:phx_new, path: "./installer", only: :test},
       {:websocket_client, git: "https://github.com/jeremyong/websocket_client.git", only: :test},
+      {:mint, "~> 1.4", only: :test},
```
it should not be necessary to mark mint as test - it's a transitive dependency from mint_web_socket which is already marked as test

electricshaman replied:
The dependency is still necessary for `Mint.HTTP` (right?), and there's a clear convention in place in phoenix with the existing test dependencies.  Are you cool with `mint` being the only dependency without the test attribute?

electricshaman replied:
Follow up: Is Mint actually a transitive dependency here if the consuming code has to connect via Mint.HTTP.connect?

the-mikedavis replied:
oh I see - no you're right, we want this as a top-level dependency since we'll be calling `Mint.HTTP.connect/4`. In this case mint is not transitive because we're calling functions within it directly :+1:

the-mikedavis committed:
remove websocket_client dependency

the-mikedavis committed:
rewrite integration-test websocket client with Mint.WebSocket

This is a rather large refactor commit which is a complete rewrite
of test/support/websocket_client.exs.

For the integration tests, there are only a handful of changes:

- `start_link/3` has been renamed to `connect/3`, and it acts the same
- `send_message/2` and `send_control_frame/3` have been replaced with
  a unified `send/2` which can send any frame
    - this only affects the tests of the websocket client itself, not
      the channel tests

The WebsocketClient fixture has been overhauled. :websocket_client
is a general-purpose websocket client library that uses processes
to hide some of the minutiae of the websocket runtime - namely
responding to pings with pongs and terminating the process when
a close frame is received.

Mint.WebSocket is functional, so it is less "a websocket client"
and more "tools for writing yourself a websocket client." As such,
the fixture is now a regular GenServer which responds to calls,
casts and info.

the-mikedavis committed:
fix expected port in transport URI connect-info extraction test

This is necessary because `:websocket_client` sets the HTTP/1
Host header in the bootstrap GET request without the port.

For example, the line 133 test in websocket_socket_test.exs
("returns params with sync request") gives the following request
with `:websocket_client`:

    GET /ws/websocket?key=value HTTP/1.1
    Host: 127.0.0.1
    Connection: Upgrade
    Sec-WebSocket-Version: 13
    Sec-WebSocket-Key: Wd76OpgxB6G3HhApUbE9ig==
    Upgrade: websocket

A run of the same test case with Mint:

    GET /ws/websocket?key=value HTTP/1.1
    host: 127.0.0.1:5907
    user-agent: mint/1.4.1
    upgrade: websocket
    connection: upgrade
    sec-websocket-version: 13
    sec-websocket-key: +fVzxGC42VwPgi2Tm/4bRA==

Note that the host header contains the port. So we update
the test to expect the extracted connection information to
contain the server's port as the client knows it.

the-mikedavis committed:
mark channels test as async

Marking this as async appears to have no affect on the test results.
These tests do not conflict with the tests of the websocket client
itself, so it seems fine to run them in parallel with other modules'
tests.

the-mikedavis commented:
closing this out in favor of the upstream pr
